File: futureagi/ee/usage/services/metering.py
"""Usage metering — pre-check before billable actions.

Usage:
    from ee.usage.services.metering import check_usage

    check = check_usage(str(org.id), "turing_large_evaluator")
    if not check.allowed:
        return Response({"error": check.reason}, status=403)

Synchronous. Redis-only (no DB queries). < 2ms.
"""

from __future__ import annotations

import os
from datetime import datetime

import structlog
from ee.usage.schemas.events import CheckResult, UpgradeCTA
from ee.usage.services.config import BillingConfig
from ee.usage.services.emitter import get_redis
from redis.exceptions import RedisError

logger = structlog.get_logger(__name__)

_lua_sha: str | None = None


def _get_lua_sha() -> str:
    """Load and cache the Lua check script SHA."""
    global _lua_sha
    if _lua_sha is None:
        script_path = os.path.join(
            os.path.dirname(__file__), "..", "lua", "check_quota.lua"
        )
        with open(script_path) as f:
            _lua_sha = get_redis().script_load(f.read())
    return _lua_sha


def _get_current_period() -> str:
    """Get current billing period as YYYY-MM string."""
    return datetime.utcnow().strftime("%Y-%m")


def _get_cached_plan(org_id: str) -> str:
    """Get org's plan from Redis cache, or fetch from DB and cache (5 min TTL)."""
    cache_key = f"plan:{org_id}"
    try:
        r = get_redis()
        cached = r.get(cache_key)
        if cached:
            return cached if isinstance(cached, str) else cached.decode()
    except RedisError:
        logger.warning("plan_cache_read_failed", org_id=org_id)
        r = None

    # Fetch from DB (one query, only on cache miss)
    from ee.usage.models.usage import OrganizationSubscription

    plan = (
        OrganizationSubscription.objects.filter(organization_id=org_id, deleted=False)
        .values_list("plan", flat=True)
        .first()
    )
    plan = plan or "free"
    if r is not None:
        try:
            r.setex(cache_key, 300, plan)  # 5 min TTL
        except RedisError:
            logger.warning("plan_cache_write_failed", org_id=org_id)
    return plan


def _get_cached_billing_status(org_id: str) -> str:
    """Get org's billing status from Redis cache, or fetch from DB and cache (5 min TTL)."""
    cache_key = f"billing_status:{org_id}"
    try:
        r = get_redis()
        cached = r.get(cache_key)
        if cached:
            return cached if isinstance(cached, str) else cached.decode()
    except RedisError:
        logger.warning("billing_status_cache_read_failed", org_id=org_id)
        r = None

    from ee.usage.models.usage import (
        OrganizationStatusChoices,
        OrganizationSubscription,
    )

    status = (
        OrganizationSubscription.objects.filter(organization_id=org_id, deleted=False)
        .values_list("status", flat=True)
        .first()
    )
    status = status or OrganizationStatusChoices.ACTIVE
    if r is not None:
        try:
            r.setex(cache_key, 300, status)
        except RedisError:
            logger.warning("billing_status_cache_write_failed", org_id=org_id)
    return status


def check_usage(org_id: str, event_type: str, amount: float = 0) -> CheckResult:
    """Pre-check: can this org perform this billable action?

    Synchronous. Redis-only. < 2ms.

    Args:
        org_id: Organization ID.
        event_type: API call type key from billing.yaml.
        amount: Override amount. If 0, uses per_call from config.

    Returns:
        CheckResult with allowed=True/False and details.
    """
    config = BillingConfig.get()

    # Resolve event_type → dimension + amount
    try:
        call_type_config = config.get_call_type(event_type)
    except KeyError:
        logger.warning(
            "check_usage_unknown_event_type", event_type=event_type, org_id=org_id
        )
        return CheckResult(
            allowed=True
        )  # Unknown type — don't block, log for investigation

    dimension = call_type_config.dimension

    # Stripe exhausted its retry window (grace period) without payment.
    # past_due stays soft — the grace period itself never blocks.
    from ee.usage.models.usage import OrganizationStatusChoices

    if _get_cached_billing_status(org_id) == OrganizationStatusChoices.UNPAID:
        return CheckResult(
            allowed=False,
            error_code="PAYMENT_REQUIRED",
            reason=(
                "Your subscription has unpaid invoices. "
                "Update your payment method to restore access."
            ),
            dimension=dimension,
        )

    # Resolve amount from config if not explicitly provided
    resolved_amount = amount
    if resolved_amount == 0:
        if call_type_config.per_call is not None:
            resolved_amount = call_type_config.per_call
        else:
            resolved_amount = 1

    # Get plan
    plan = _get_cached_plan(org_id)
    plan_config = config.get_plan(plan)
    period = _get_current_period()

    # Check free tier hard cap
    if plan_config.usage_caps == "hard":
        dim_config = config.get_dimension(dimension)
        free_allowance = config.get_free_allowance(dimension, plan)
        free_allowance_native = float(
            free_allowance * dim_config.native_to_display_divisor
        )

        if free_allowance_native > 0:
            usage_key = f"usage:{org_id}:{dimension}:{period}"
            result = get_redis().evalsha(
                _get_lua_sha(),
                1,
                usage_key,
                str(free_allowance_native),
                str(resolved_amount),
            )
            current = float(result)

            if current == -1:
                # Over limit
                actual_current = float(get_redis().get(usage_key) or 0)
                return CheckResult(
                    allowed=False,
                    error_code="FREE_TIER_LIMIT",
                    reason=(
                        f"Free tier {dim_config.display_name} limit reached "
                        f"({actual_current:,.2f} / {free_allowance_native:,.2f} {dim_config.native_unit})"
                    ),
                    dimension=dimension,
                    current_usage=actual_current,
                    limit=free_allowance_native,
                    upgrade_cta=UpgradeCTA(
                        text="Upgrade to Pay-as-you-go for unlimited usage",
                        plan="payg",
                    ),
                )

    # Check budget pause flag (set by budget enforcement in Phase 4)
    pause_key = f"pause:{org_id}:{dimension}"
    if get_redis().get(pause_key):
        return CheckResult(
            allowed=False,
            error_code="BUDGET_PAUSED",
            reason=f"Usage paused — you set a budget limit for {dimension}",
            dimension=dimension,
        )

    return CheckResult(allowed=True, dimension=dimension)


QUOTA_KEY_TTL = 3600


def publish_quota(org_id: str, dimension: str, limit_native: float) -> None:
    r = get_redis()
    key = f"quota:{org_id}:{dimension}"
    r.setex(key, QUOTA_KEY_TTL, str(limit_native))


def publish_quotas_for_org(org_id: str) -> None:
    config = BillingConfig.get()
    plan = _get_cached_plan(org_id)
    plan_config = config.get_plan(plan)

    for dim_key, dim_config in config.get_all_dimensions().items():
        if plan_config.usage_caps == "hard":
            free_allowance = config.get_free_allowance(dim_key, plan)
            limit_native = float(free_allowance * dim_config.native_to_display_divisor)
        else:
            limit_native = -1

        publish_quota(org_id, dim_key, limit_native)

File: fi-collector/pkg/auth/metering.go
package auth

import (
	"context"
	"fmt"
	"log/slog"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/redis/go-redis/v9"
)

var checkQuotaLua = `
local key = KEYS[1]
local limit = tonumber(ARGV[1])
local amount = tonumber(ARGV[2]) or 1
local current = tonumber(redis.call('GET', key) or '0')
if limit <= 0 then return current end
if (current + amount) > limit then return -1 end
return current
`

var freeTierAllowances = map[string]int64{
	"tracing_events": 50_000,
	"storage":        50 * 1024 * 1024 * 1024,
}

var hardCapPlans = map[string]bool{
	"free": true,
}

var eventTypeToDimension = map[string]string{
	"tracing_event": "tracing_events",
	"observe_add":   "storage",
}

// CheckResult holds the outcome of a quota check.
type CheckResult struct {
	Allowed   bool
	Reason    string
	ErrorCode string
	Dimension string
}

// Metering enforces pre-check quota limits before ingestion.
type Metering struct {
	rdb    *redis.Client
	pg     *pgxpool.Pool
	log    *slog.Logger
	luaSHA string
}

// NewMetering creates a metering instance. Returns nil if rdb is nil.
func NewMetering(rdb *redis.Client, pgRead *pgxpool.Pool, log *slog.Logger) *Metering {
	if rdb == nil {
		return nil
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	sha, err := rdb.ScriptLoad(ctx, checkQuotaLua).Result()
	if err != nil {
		log.Warn("metering lua script load failed, quota checks disabled", "err", err)
		return &Metering{rdb: rdb, pg: pgRead, log: log}
	}

	return &Metering{rdb: rdb, pg: pgRead, log: log, luaSHA: sha}
}

// CheckUsage checks if an org can perform a billable action.
// Fail-open: returns Allowed=true on any error.
func (m *Metering) CheckUsage(ctx context.Context, orgID, eventType string, amount int64) CheckResult {
	if m == nil {
		return CheckResult{Allowed: true}
	}

	dimension, ok := eventTypeToDimension[eventType]
	if !ok {
		m.log.Warn("unknown event type in check_usage", "event_type", eventType)
		return CheckResult{Allowed: true}
	}

	plan := m.getCachedPlan(ctx, orgID)

	if hardCapPlans[plan] {
		allowance, hasAllowance := freeTierAllowances[dimension]
		if hasAllowance && allowance > 0 && m.luaSHA != "" {
			period := time.Now().UTC().Format("2006-01")
			usageKey := fmt.Sprintf("usage:%s:%s:%s", orgID, dimension, period)

			result, err := m.rdb.EvalSha(ctx, m.luaSHA, []string{usageKey},
				fmt.Sprintf("%d", allowance),
				fmt.Sprintf("%d", amount),
			).Int64()
			if err != nil {
				m.log.Warn("quota check lua failed, allowing", "err", err, "org", orgID)
				return CheckResult{Allowed: true, Dimension: dimension}
			}
			if result == -1 {
				return CheckResult{
					Allowed:   false,
					ErrorCode: "FREE_TIER_LIMIT",
					Reason:    fmt.Sprintf("Free tier %s limit reached", dimension),
					Dimension: dimension,
				}
			}
		}
	}

	// Check budget pause flag
	pauseKey := fmt.Sprintf("pause:%s:%s", orgID, dimension)
	paused, err := m.rdb.Get(ctx, pauseKey).Result()
	if err == nil && paused != "" {
		return CheckResult{
			Allowed:   false,
			ErrorCode: "BUDGET_PAUSED",
			Reason:    fmt.Sprintf("Usage paused — budget limit set for %s", dimension),
			Dimension: dimension,
		}
	}

	return CheckResult{Allowed: true, Dimension: dimension}
}

// getCachedPlan returns the org's billing plan from Redis cache, falling back to PG.
func (m *Metering) getCachedPlan(ctx context.Context, orgID string) string {
	cacheKey := fmt.Sprintf("plan:%s", orgID)

	cached, err := m.rdb.Get(ctx, cacheKey).Result()
	if err == nil && cached != "" {
		return cached
	}

	if m.pg == nil {
		return "free"
	}

	const q = `SELECT plan FROM usage_organizationsubscription
		WHERE organization_id = $1 AND deleted = false LIMIT 1`

	var plan string
	err = m.pg.QueryRow(ctx, q, orgID).Scan(&plan)
	if err != nil {
		if err != pgx.ErrNoRows {
			m.log.Warn("plan lookup failed", "err", err, "org", orgID)
		}
		plan = "free"
	}

	_ = m.rdb.SetEx(ctx, cacheKey, plan, 5*time.Minute).Err()

	return plan
}
